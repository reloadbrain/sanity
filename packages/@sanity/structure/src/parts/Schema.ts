import getDefaultModule from './getDefaultModule'

interface Schema {
  name: string
  get(typeName: string): any
  getTypeNames(): string[]
}

// We are lazy-loading the part to work around typescript trying to resolve it
const defaultSchema = ((): Schema => {
  const schema: Schema = getDefaultModule(require('part:@sanity/base/schema'))
  return schema
})()

export {Schema, defaultSchema}
