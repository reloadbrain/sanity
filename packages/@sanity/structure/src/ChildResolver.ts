import {StructureNode, EditorNode} from './StructureNodes'

export interface ChildResolverOptions {
  index: number
}

export interface ChildResolver {
  (itemId: string, parent: StructureNode, options: ChildResolverOptions):
    | StructureNode
    | EditorNode
    | Promise<StructureNode | EditorNode>
    | undefined
}
