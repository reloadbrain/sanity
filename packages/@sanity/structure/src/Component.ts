import {SerializeOptions, StructureNode, Serializable} from './StructureNodes'
import {SerializeError, HELP_URL} from './SerializeError'
import {MenuItem, MenuItemBuilder, maybeSerializeMenuItem} from './MenuItem'
import {MenuItemGroup, MenuItemGroupBuilder, maybeSerializeMenuItemGroup} from './MenuItemGroup'

export interface Component extends StructureNode {
  component: Function
  menuItems: MenuItem[]
  menuItemGroups: MenuItemGroup[]
}

export interface ComponentInput extends StructureNode {
  component: Function
  menuItems?: (MenuItem | MenuItemBuilder)[]
  menuItemGroups?: (MenuItemGroup | MenuItemGroupBuilder)[]
}

export interface BuildableList extends Partial<StructureNode> {
  component?: Function
  menuItems?: (MenuItem | MenuItemBuilder)[]
  menuItemGroups?: (MenuItemGroup | MenuItemGroupBuilder)[]
}

export class ComponentBuilder implements Serializable {
  protected spec: BuildableList

  constructor(spec?: ComponentInput) {
    this.spec = spec ? spec : {}
  }

  id(id: string) {
    this.spec.id = id
    return this
  }

  title(title: string) {
    this.spec.title = title
    return this
  }

  component(component: Function) {
    this.spec.component = component
    return this
  }

  menuItems(items: (MenuItem | MenuItemBuilder)[]) {
    this.spec.menuItems = items
    return this
  }

  menuItemGroups(groups: (MenuItemGroup | MenuItemGroupBuilder)[]) {
    this.spec.menuItemGroups = groups
    return this
  }

  serialize(options: SerializeOptions = {path: []}): Component {
    const {id, title, component} = this.spec
    if (typeof id !== 'string' || !id) {
      throw new SerializeError(
        '`id` is required for `component` structure item',
        options.path,
        options.index
      ).withHelpUrl(HELP_URL.ID_REQUIRED)
    }

    if (!component) {
      throw new SerializeError(
        '`component` is required for `component` structure item',
        options.path,
        options.index
      ).withHelpUrl(HELP_URL.ID_REQUIRED)
    }

    return {
      id,
      title,
      type: 'component',
      component,
      menuItems: (this.spec.menuItems || []).map((item, i) =>
        maybeSerializeMenuItem(item, i, options.path)
      ),
      menuItemGroups: (this.spec.menuItemGroups || []).map((item, i) =>
        maybeSerializeMenuItemGroup(item, i, options.path)
      )
    }
  }
}
