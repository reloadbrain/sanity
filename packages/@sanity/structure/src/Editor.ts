import {EditorNode} from './StructureNodes'

export type PartialEditorNode = {
  id?: string
  title?: string
  options?: {
    id?: string
    type?: string
  }
}

export class EditorBuilder {
  protected spec: PartialEditorNode

  constructor(spec?: PartialEditorNode) {
    this.spec = spec || {}
  }

  id(id: string): EditorBuilder {
    this.spec.id = id
    return this
  }

  title(title: string): EditorBuilder {
    this.spec.title = title
    return this
  }

  documentId(documentId: string): EditorBuilder {
    this.spec.options = {
      ...(this.spec.options || {}),
      id: documentId
    }

    return this
  }

  type(documentType: string): EditorBuilder {
    this.spec.options = {
      ...(this.spec.options || {}),
      type: documentType
    }

    return this
  }

  serialize(): EditorNode {
    const {id, options} = this.spec
    if (typeof id !== 'string' || !id) {
      throw new Error('`id` is required for editor nodes')
    }

    if (!options || !options.id) {
      throw new Error('document id (`id`) is required for editor nodes')
    }

    if (!options || !options.type) {
      throw new Error('document type (`type`) is required for editor nodes')
    }

    return {
      ...this.spec,
      id,
      type: 'document',
      options: {id: options.id, type: options.type}
    }
  }
}
