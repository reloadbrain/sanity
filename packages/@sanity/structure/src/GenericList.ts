import {StructureNode} from './StructureNodes'
import {ChildResolver} from './ChildResolver'
import {Layout, LayoutOptions} from './Layout'
import {MenuItem, MenuItemBuilder} from './MenuItem'
import {MenuItemGroup, MenuItemGroupBuilder} from './MenuItemGroup'
import {IntentChecker} from './Intent'

function maybeSerializeMenuItemGroup(item: MenuItemGroup | MenuItemGroupBuilder): MenuItemGroup {
  return item instanceof MenuItemGroupBuilder ? item.serialize() : item
}

function maybeSerializeMenuItem(item: MenuItem | MenuItemBuilder): MenuItem {
  return item instanceof MenuItemBuilder ? item.serialize() : item
}

function noChildResolver() {
  return undefined
}

export interface GenericList extends StructureNode {
  menuItems: MenuItem[]
  menuItemGroups: MenuItemGroup[]
  defaultLayout?: Layout
  canHandleIntent?: IntentChecker
  resolveChildForItem: ChildResolver
}

export type PartialGenericList = Partial<GenericList>

export abstract class GenericListBuilder<L extends PartialGenericList> {
  protected intentChecker?: IntentChecker

  constructor(protected spec: L) {}

  id(id: string) {
    this.spec.id = id
    return this
  }

  title(title: string) {
    this.spec.title = title
    return this
  }

  layout(layout: Layout) {
    if (!LayoutOptions.includes(layout)) {
      throw new Error(
        `\`layout\` must be one of ${LayoutOptions.map(item => `"${item}"`).join(', ')}`
      )
    }

    this.spec.defaultLayout = layout
    return this
  }

  menuItems(items: (MenuItem | MenuItemBuilder)[]) {
    this.spec.menuItems = items.map(maybeSerializeMenuItem)
    return this
  }

  menuItemGroups(groups: (MenuItemGroup | MenuItemGroupBuilder)[]) {
    this.spec.menuItemGroups = groups.map(maybeSerializeMenuItemGroup)
    return this
  }

  childResolver(resolver: ChildResolver) {
    this.spec.resolveChildForItem = resolver
    return this
  }

  canHandleIntent(checker: IntentChecker) {
    this.intentChecker = checker
    return this
  }

  serialize(): GenericList {
    return {
      id: this.spec.id || '',
      title: this.spec.title,
      type: 'genericList',
      defaultLayout: this.spec.defaultLayout,
      resolveChildForItem: this.spec.resolveChildForItem || noChildResolver,
      canHandleIntent: this.intentChecker,
      menuItems: this.spec.menuItems || [],
      menuItemGroups: this.spec.menuItemGroups || []
    }
  }
}
