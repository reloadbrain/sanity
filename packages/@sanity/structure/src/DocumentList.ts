import {StructureNode, EditorNode} from './StructureNodes'
import {GenericListBuilder, PartialGenericList, GenericList} from './GenericList'
import {ChildResolver, ChildResolverOptions} from './ChildResolver'
import {SortItem} from './Sort'

const resolveChildForItem: ChildResolver = (
  itemId: string,
  parent: StructureNode,
  options: ChildResolverOptions
): StructureNode | EditorNode | Promise<StructureNode | EditorNode> | undefined => {
  const parentItem = parent as DocumentList
  return {
    id: 'editor',
    type: 'document',
    options: {
      id: itemId,
      // @todo this aint always right
      type: parentItem.options.params && parentItem.options.params.type
    }
  }
}

export interface PartialDocumentList extends PartialGenericList {
  options?: DocumentListOptions
}

export interface DocumentList extends GenericList {
  options: DocumentListOptions
  resolveChildForItem: ChildResolver
}

interface DocumentListOptions {
  filter: string
  params?: {[key: string]: any}
  defaultOrdering?: SortItem[]
}

export class DocumentListBuilder extends GenericListBuilder<PartialDocumentList> {
  constructor(spec: PartialDocumentList = {}) {
    super(spec)
  }

  filter(filter: string): DocumentListBuilder {
    this.spec.options = {...(this.spec.options || {}), filter}
    return this
  }

  params(params: {}): DocumentListBuilder {
    this.spec.options = {...(this.spec.options || {filter: ''}), params}
    return this
  }

  defaultOrdering(ordering: SortItem[]): DocumentListBuilder {
    if (!Array.isArray(ordering)) {
      throw new Error('`defaultOrdering` must be an array of order clauses')
    }

    this.spec.options = {...(this.spec.options || {filter: ''}), defaultOrdering: ordering}
    return this
  }

  serialize(): DocumentList {
    if (typeof this.spec.id !== 'string' || !this.spec.id) {
      throw new Error('`id` is required for document lists')
    }

    if (!this.spec.options || !this.spec.options.filter) {
      throw new Error('`filter` is required for document lists')
    }

    return {
      ...super.serialize(),
      type: 'documentList',
      resolveChildForItem: this.spec.resolveChildForItem || resolveChildForItem,
      options: this.spec.options
    }
  }
}
