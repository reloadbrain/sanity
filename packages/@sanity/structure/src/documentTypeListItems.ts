import {Schema, defaultSchema} from './parts/schema'
import {dataAspects} from './parts/DataAspects'
import {ListItem, ListItemBuilder, SchemaType} from './ListItem'
import {DocumentList, DocumentListBuilder} from './DocumentList'
import {MenuItemBuilder} from './MenuItem'
import {DEFAULT_ORDERING_OPTIONS} from './Sort'
import {getPlusIcon, getSortIcon, getListIcon} from './parts/icon'
import {EditorBuilder} from './Editor'

const PlusIcon = getPlusIcon()
const SortIcon = getSortIcon()
const ListIcon = getListIcon()

export const getDocumentTypeListItems = (schema: Schema = defaultSchema): ListItem[] => {
  const resolver = dataAspects(schema)
  const types = resolver.getInferredTypes()

  return types.map(name => {
    const title = resolver.getDisplayName(name)
    const schemaType = schema.get(name)
    return getDocumentTypeListItem(name, title, schemaType)
  })
}

function getDocumentTypeListItem(name: string, title: string, type: SchemaType): ListItem {
  return new ListItemBuilder()
    .id(name)
    .title(title)
    .schemaType(type)
    .child(getDocumentList(name, title, type))
    .serialize()
}

function getDocumentList(name: string, title: string, type: SchemaType): DocumentList {
  return new DocumentListBuilder()
    .id(name)
    .title(title)
    .filter('_type == $type')
    .params({type: name})
    .defaultOrdering([]) // @todo
    .canHandleIntent(
      (intentName, params): boolean =>
        Boolean(intentName === 'edit' && params && params.id) ||
        Boolean(intentName === 'create' && params && params.type === name)
    )
    .menuItemGroups([
      {id: 'sorting', title: 'Sort'},
      {id: 'layout', title: 'Layout'},
      {id: 'actions', title: 'Actions'}
    ])
    .childResolver((documentId: string) =>
      new EditorBuilder()
        .id('editor')
        .type(name)
        .documentId(documentId)
        .serialize()
    )
    .menuItems([
      // Create new (from action button)
      new MenuItemBuilder()
        .title(`Create new ${title}`)
        .icon(PlusIcon)
        .intent({type: 'create', params: {type: name}})
        .showAsAction(true),

      // Sort by <Y>
      ...getOrderingMenuItemsForSchemaType(type),

      // Display as <Z>
      new MenuItemBuilder()
        .group('layout')
        .title('List')
        .icon(ListIcon)
        .action('setLayout')
        .params({layout: 'default'}),

      new MenuItemBuilder()
        .group('layout')
        .title('Details')
        .icon(ListIcon)
        .action('setLayout')
        .params({layout: 'detail'}),

      // Create new (from menu)
      new MenuItemBuilder()
        .group('actions')
        .title('Create newâ€¦')
        .icon(PlusIcon)
        .intent({type: 'create', params: {type: name}})
    ])
    .serialize()
}

function getOrderingMenuItemsForSchemaType(type: SchemaType) {
  return (type.orderings
    ? type.orderings.concat(DEFAULT_ORDERING_OPTIONS)
    : DEFAULT_ORDERING_OPTIONS
  ).map(ordering =>
    new MenuItemBuilder()
      .group('sorting')
      .title(`Sort by ${ordering.title}`)
      .icon(SortIcon)
      .action('setSortOrder')
      .params({by: ordering.by})
  )
}
