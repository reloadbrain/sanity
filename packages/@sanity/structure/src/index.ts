import {ListBuilder, ListInput} from './List'
import {getDocumentTypeListItems} from './documentTypeListItems'
import {MenuItemBuilder, MenuItem} from './MenuItem'
import {ListItemBuilder, ListItem} from './ListItem'
import {MenuItemGroup, MenuItemGroupBuilder} from './MenuItemGroup'
import {DocumentListBuilder, DocumentListInput} from './DocumentList'
import {EditorBuilder} from './Editor'
import {EditorNode} from './StructureNodes'

const StructureBuilder = {
  list: (spec?: ListInput): ListBuilder => new ListBuilder(spec),
  listItem: (spec?: ListItem): ListItemBuilder => new ListItemBuilder(spec),
  editor: (spec?: EditorNode): EditorBuilder => new EditorBuilder(spec),
  menuItem: (spec?: MenuItem): MenuItemBuilder => new MenuItemBuilder(spec),
  menuItemGroup: (spec?: MenuItemGroup): MenuItemGroupBuilder => new MenuItemGroupBuilder(spec),
  documentList: (spec?: DocumentListInput): DocumentListBuilder => new DocumentListBuilder(spec),
  documentTypeListItems: getDocumentTypeListItems
}

export {StructureBuilder}
