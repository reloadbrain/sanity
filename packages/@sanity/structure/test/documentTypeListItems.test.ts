import {StructureBuilder as S} from '../src'
import {defaultSchema} from '../src/parts/Schema'
import {DocumentList} from '../src/DocumentList'
import serializeStructure from './util/serializeStructure'

test('generates correct document type list items from global schema', () => {
  expect(S.documentTypeListItems()).toMatchSnapshot()
})

test('generates correct document type list items from given schema', () => {
  expect(S.documentTypeListItems(defaultSchema)).toMatchSnapshot()
})

test('generates correct document type list item for specific type', () => {
  expect(S.documentTypeListItem('author')).toMatchSnapshot()
  expect(S.documentTypeListItem('author', defaultSchema)).toMatchSnapshot()
})

test('generated canHandleIntent responds to edit/create on document type', () => {
  const listItems = S.documentTypeListItems().map(item => item.serialize())
  expect(listItems).toHaveLength(2)
  expect(listItems[0]).toMatchObject({id: 'author', title: 'Author'})
  expect(listItems[0]).toHaveProperty('child')
  expect(listItems[0].child).toHaveProperty('canHandleIntent')

  const child = listItems[0].child as DocumentList
  expect(child.canHandleIntent('create', {type: 'book'})).toBe(false)
  expect(child.canHandleIntent('create', {type: 'author'})).toBe(true)
  expect(child.canHandleIntent('edit', {id: 'wow'})).toBe(true)
})

test('generated document panes responds with correct editor child', done => {
  const listItems = S.documentTypeListItems().map(item => item.serialize())
  expect(listItems).toHaveLength(2)
  expect(listItems[0]).toMatchObject({id: 'author', title: 'Author'})
  expect(listItems[0]).toHaveProperty('child')

  const listItem = listItems[0].child as DocumentList
  const context = {parent: listItem, index: 0}
  serializeStructure(listItem.child, context, ['grrm', context]).subscribe(itemChild => {
    expect(itemChild).toMatchObject({
      id: 'editor',
      options: {
        type: 'author',
        id: 'grrm'
      }
    })

    done()
  })
})
