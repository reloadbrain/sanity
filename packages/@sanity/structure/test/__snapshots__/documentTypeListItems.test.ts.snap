// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates correct document type list items from given schema 1`] = `
Array [
  Object {
    "child": Object {
      "canHandleIntent": [Function],
      "defaultLayout": undefined,
      "id": "author",
      "menuItemGroups": Array [
        Object {
          "id": "sorting",
          "title": "Sort",
        },
        Object {
          "id": "layout",
          "title": "Layout",
        },
        Object {
          "id": "actions",
          "title": "Actions",
        },
      ],
      "menuItems": Array [
        Object {
          "icon": [Function],
          "intent": Object {
            "params": Object {
              "type": "author",
            },
            "type": "create",
          },
          "showAsAction": true,
          "title": "Create new Author",
        },
        Object {
          "action": "setSortOrder",
          "group": "sorting",
          "icon": [Function],
          "params": Object {
            "by": Array [
              Object {
                "direction": "desc",
                "field": "_updatedAt",
              },
            ],
          },
          "title": "Sort by Last edited",
        },
        Object {
          "action": "setSortOrder",
          "group": "sorting",
          "icon": [Function],
          "params": Object {
            "by": Array [
              Object {
                "direction": "desc",
                "field": "_createdAt",
              },
            ],
          },
          "title": "Sort by Created",
        },
        Object {
          "action": "setLayout",
          "group": "layout",
          "icon": [Function],
          "params": Object {
            "layout": "default",
          },
          "title": "List",
        },
        Object {
          "action": "setLayout",
          "group": "layout",
          "icon": [Function],
          "params": Object {
            "layout": "detail",
          },
          "title": "Details",
        },
        Object {
          "group": "actions",
          "icon": [Function],
          "intent": Object {
            "params": Object {
              "type": "author",
            },
            "type": "create",
          },
          "title": "Create new…",
        },
      ],
      "options": Object {
        "defaultOrdering": Array [],
        "filter": "_type == $type",
        "params": Object {
          "type": "author",
        },
      },
      "resolveChildForItem": [Function],
      "title": "Author",
      "type": "documentList",
    },
    "id": "author",
    "schemaType": Object {
      "fields": Array [
        Object {
          "name": "name",
          "title": "Name",
          "type": "string",
        },
        Object {
          "name": "slug",
          "options": Object {
            "maxLength": 96,
            "source": "name",
          },
          "title": "Slug",
          "type": "slug",
        },
        Object {
          "name": "image",
          "options": Object {
            "hotspot": true,
          },
          "title": "Image",
          "type": "image",
        },
      ],
      "name": "author",
      "preview": Object {
        "select": Object {
          "media": "image",
          "title": "name",
        },
      },
      "title": "Author",
      "type": "document",
    },
    "title": "Author",
  },
  Object {
    "child": Object {
      "canHandleIntent": [Function],
      "defaultLayout": undefined,
      "id": "post",
      "menuItemGroups": Array [
        Object {
          "id": "sorting",
          "title": "Sort",
        },
        Object {
          "id": "layout",
          "title": "Layout",
        },
        Object {
          "id": "actions",
          "title": "Actions",
        },
      ],
      "menuItems": Array [
        Object {
          "icon": [Function],
          "intent": Object {
            "params": Object {
              "type": "post",
            },
            "type": "create",
          },
          "showAsAction": true,
          "title": "Create new Post",
        },
        Object {
          "action": "setSortOrder",
          "group": "sorting",
          "icon": [Function],
          "params": Object {
            "by": Array [
              Object {
                "direction": "asc",
                "field": "title",
              },
              Object {
                "direction": "asc",
                "field": "publicationYear",
              },
            ],
          },
          "title": "Sort by Title",
        },
        Object {
          "action": "setSortOrder",
          "group": "sorting",
          "icon": [Function],
          "params": Object {
            "by": Array [
              Object {
                "direction": "asc",
                "field": "translations.se",
              },
              Object {
                "direction": "asc",
                "field": "title",
              },
            ],
          },
          "title": "Sort by Swedish title",
        },
        Object {
          "action": "setSortOrder",
          "group": "sorting",
          "icon": [Function],
          "params": Object {
            "by": Array [
              Object {
                "direction": "desc",
                "field": "_updatedAt",
              },
            ],
          },
          "title": "Sort by Last edited",
        },
        Object {
          "action": "setSortOrder",
          "group": "sorting",
          "icon": [Function],
          "params": Object {
            "by": Array [
              Object {
                "direction": "desc",
                "field": "_createdAt",
              },
            ],
          },
          "title": "Sort by Created",
        },
        Object {
          "action": "setLayout",
          "group": "layout",
          "icon": [Function],
          "params": Object {
            "layout": "default",
          },
          "title": "List",
        },
        Object {
          "action": "setLayout",
          "group": "layout",
          "icon": [Function],
          "params": Object {
            "layout": "detail",
          },
          "title": "Details",
        },
        Object {
          "group": "actions",
          "icon": [Function],
          "intent": Object {
            "params": Object {
              "type": "post",
            },
            "type": "create",
          },
          "title": "Create new…",
        },
      ],
      "options": Object {
        "defaultOrdering": Array [],
        "filter": "_type == $type",
        "params": Object {
          "type": "post",
        },
      },
      "resolveChildForItem": [Function],
      "title": "Post",
      "type": "documentList",
    },
    "id": "post",
    "schemaType": Object {
      "fields": Array [
        Object {
          "name": "title",
          "title": "Title",
          "type": "string",
        },
        Object {
          "name": "slug",
          "options": Object {
            "maxLength": 96,
            "source": "title",
          },
          "title": "Slug",
          "type": "slug",
        },
        Object {
          "name": "author",
          "title": "Author",
          "to": Object {
            "type": "author",
          },
          "type": "reference",
        },
        Object {
          "name": "mainImage",
          "options": Object {
            "hotspot": true,
          },
          "title": "Main image",
          "type": "image",
        },
        Object {
          "name": "publishedAt",
          "title": "Published at",
          "type": "datetime",
        },
        Object {
          "name": "body",
          "title": "Body",
          "type": "text",
        },
      ],
      "name": "post",
      "orderings": Array [
        Object {
          "by": Array [
            Object {
              "direction": "asc",
              "field": "title",
            },
            Object {
              "direction": "asc",
              "field": "publicationYear",
            },
          ],
          "name": "title",
          "title": "Title",
        },
        Object {
          "by": Array [
            Object {
              "direction": "asc",
              "field": "translations.se",
            },
            Object {
              "direction": "asc",
              "field": "title",
            },
          ],
          "name": "swedishTitle",
          "title": "Swedish title",
        },
      ],
      "preview": Object {
        "select": Object {
          "author": "author.name",
          "media": "mainImage",
          "title": "title",
        },
      },
      "title": "Post",
      "type": "document",
    },
    "title": "Post",
  },
]
`;

exports[`generates correct document type list items from global schema 1`] = `
Array [
  Object {
    "child": Object {
      "canHandleIntent": [Function],
      "defaultLayout": undefined,
      "id": "author",
      "menuItemGroups": Array [
        Object {
          "id": "sorting",
          "title": "Sort",
        },
        Object {
          "id": "layout",
          "title": "Layout",
        },
        Object {
          "id": "actions",
          "title": "Actions",
        },
      ],
      "menuItems": Array [
        Object {
          "icon": [Function],
          "intent": Object {
            "params": Object {
              "type": "author",
            },
            "type": "create",
          },
          "showAsAction": true,
          "title": "Create new Author",
        },
        Object {
          "action": "setSortOrder",
          "group": "sorting",
          "icon": [Function],
          "params": Object {
            "by": Array [
              Object {
                "direction": "desc",
                "field": "_updatedAt",
              },
            ],
          },
          "title": "Sort by Last edited",
        },
        Object {
          "action": "setSortOrder",
          "group": "sorting",
          "icon": [Function],
          "params": Object {
            "by": Array [
              Object {
                "direction": "desc",
                "field": "_createdAt",
              },
            ],
          },
          "title": "Sort by Created",
        },
        Object {
          "action": "setLayout",
          "group": "layout",
          "icon": [Function],
          "params": Object {
            "layout": "default",
          },
          "title": "List",
        },
        Object {
          "action": "setLayout",
          "group": "layout",
          "icon": [Function],
          "params": Object {
            "layout": "detail",
          },
          "title": "Details",
        },
        Object {
          "group": "actions",
          "icon": [Function],
          "intent": Object {
            "params": Object {
              "type": "author",
            },
            "type": "create",
          },
          "title": "Create new…",
        },
      ],
      "options": Object {
        "defaultOrdering": Array [],
        "filter": "_type == $type",
        "params": Object {
          "type": "author",
        },
      },
      "resolveChildForItem": [Function],
      "title": "Author",
      "type": "documentList",
    },
    "id": "author",
    "schemaType": Object {
      "fields": Array [
        Object {
          "name": "name",
          "title": "Name",
          "type": "string",
        },
        Object {
          "name": "slug",
          "options": Object {
            "maxLength": 96,
            "source": "name",
          },
          "title": "Slug",
          "type": "slug",
        },
        Object {
          "name": "image",
          "options": Object {
            "hotspot": true,
          },
          "title": "Image",
          "type": "image",
        },
      ],
      "name": "author",
      "preview": Object {
        "select": Object {
          "media": "image",
          "title": "name",
        },
      },
      "title": "Author",
      "type": "document",
    },
    "title": "Author",
  },
  Object {
    "child": Object {
      "canHandleIntent": [Function],
      "defaultLayout": undefined,
      "id": "post",
      "menuItemGroups": Array [
        Object {
          "id": "sorting",
          "title": "Sort",
        },
        Object {
          "id": "layout",
          "title": "Layout",
        },
        Object {
          "id": "actions",
          "title": "Actions",
        },
      ],
      "menuItems": Array [
        Object {
          "icon": [Function],
          "intent": Object {
            "params": Object {
              "type": "post",
            },
            "type": "create",
          },
          "showAsAction": true,
          "title": "Create new Post",
        },
        Object {
          "action": "setSortOrder",
          "group": "sorting",
          "icon": [Function],
          "params": Object {
            "by": Array [
              Object {
                "direction": "asc",
                "field": "title",
              },
              Object {
                "direction": "asc",
                "field": "publicationYear",
              },
            ],
          },
          "title": "Sort by Title",
        },
        Object {
          "action": "setSortOrder",
          "group": "sorting",
          "icon": [Function],
          "params": Object {
            "by": Array [
              Object {
                "direction": "asc",
                "field": "translations.se",
              },
              Object {
                "direction": "asc",
                "field": "title",
              },
            ],
          },
          "title": "Sort by Swedish title",
        },
        Object {
          "action": "setSortOrder",
          "group": "sorting",
          "icon": [Function],
          "params": Object {
            "by": Array [
              Object {
                "direction": "desc",
                "field": "_updatedAt",
              },
            ],
          },
          "title": "Sort by Last edited",
        },
        Object {
          "action": "setSortOrder",
          "group": "sorting",
          "icon": [Function],
          "params": Object {
            "by": Array [
              Object {
                "direction": "desc",
                "field": "_createdAt",
              },
            ],
          },
          "title": "Sort by Created",
        },
        Object {
          "action": "setLayout",
          "group": "layout",
          "icon": [Function],
          "params": Object {
            "layout": "default",
          },
          "title": "List",
        },
        Object {
          "action": "setLayout",
          "group": "layout",
          "icon": [Function],
          "params": Object {
            "layout": "detail",
          },
          "title": "Details",
        },
        Object {
          "group": "actions",
          "icon": [Function],
          "intent": Object {
            "params": Object {
              "type": "post",
            },
            "type": "create",
          },
          "title": "Create new…",
        },
      ],
      "options": Object {
        "defaultOrdering": Array [],
        "filter": "_type == $type",
        "params": Object {
          "type": "post",
        },
      },
      "resolveChildForItem": [Function],
      "title": "Post",
      "type": "documentList",
    },
    "id": "post",
    "schemaType": Object {
      "fields": Array [
        Object {
          "name": "title",
          "title": "Title",
          "type": "string",
        },
        Object {
          "name": "slug",
          "options": Object {
            "maxLength": 96,
            "source": "title",
          },
          "title": "Slug",
          "type": "slug",
        },
        Object {
          "name": "author",
          "title": "Author",
          "to": Object {
            "type": "author",
          },
          "type": "reference",
        },
        Object {
          "name": "mainImage",
          "options": Object {
            "hotspot": true,
          },
          "title": "Main image",
          "type": "image",
        },
        Object {
          "name": "publishedAt",
          "title": "Published at",
          "type": "datetime",
        },
        Object {
          "name": "body",
          "title": "Body",
          "type": "text",
        },
      ],
      "name": "post",
      "orderings": Array [
        Object {
          "by": Array [
            Object {
              "direction": "asc",
              "field": "title",
            },
            Object {
              "direction": "asc",
              "field": "publicationYear",
            },
          ],
          "name": "title",
          "title": "Title",
        },
        Object {
          "by": Array [
            Object {
              "direction": "asc",
              "field": "translations.se",
            },
            Object {
              "direction": "asc",
              "field": "title",
            },
          ],
          "name": "swedishTitle",
          "title": "Swedish title",
        },
      ],
      "preview": Object {
        "select": Object {
          "author": "author.name",
          "media": "mainImage",
          "title": "title",
        },
      },
      "title": "Post",
      "type": "document",
    },
    "title": "Post",
  },
]
`;
